docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+enables them to evolve and improve without limitations. The IBC module may
+not be affected by some upgrades, but some may require it to be upgraded as well
+to maintain high-value connections to other chains secure. For having this
+capability, chains that implement `IBC-rs` may bring various concerns and
+characteristics than Tendermint chains leading to different ways for upgrading
+their clients. However there are general rules that apply to all and can serve
+as a framework for any `IBC-rs` implementation. On this basis, this record aims
+to justify the logic behind upgrading a light client, determine the boundary
+between client-wide and client-specific upgrade processes, list requisites for
+validation and execution steps, and explain Tendermint's upgrade client
+implementation within the
+[ics07_tendermint](../../crates/ibc/src/clients/ics07_tendermint).
+
+## Decision
+
+In this section, after presenting rules that mainly derived from the IBC
+protocol and the review of the `IBC-Go` implementation, the upgrade client
+process for Tendermint chains is outlined in detail. Then, the design and
+implementation of this rationale in `IBC-rs` is explained.
+
+### Chain-wide Upgrade Rules
+
+#### Chain supports IBC upgrades?
+
+* IBC currently **ONLY** supports planned upgrades that are committed to in
+  advance by the upgrading chain, in order for counterparty clients to maintain
+  their connections securely.
+* An IBC upgrade **MUST** be performed if upgrading an IBC-connected chain
+  breaks the counterparty IBC client.
+* There **MUST** be a proof verification process to check upgraded client and
+  consensus states against the host chain's state.
+* Chain upgrades **MUST NOT** result in changing
+  [ClientState](../../crates/ibc/src/core/ics02_client/client_state.rs#ClientState)
+  or
+  [ConsensusState](../../crates/ibc/src/core/ics02_client/consensus_state.rs#ConsensusState)
+  implementations
+* It is **UP TO** the chain's architecture how updated client and consensus
+  states are committed, either through decentralized approaches, like governance
+  or centralized methods, like a multisig account, etc.
+* Upon commitment, chain's store **MUST** contain updated client and consensus
+  states, which can be retrieved using respective upgrade paths
+
+#### IBC Handler accepts upgrade?
+
+After ensuring that the chain upgrades is supported by IBC, the general
+validation and execution steps that apply to all chains are as follows. The
+criteria for classifying a validation as client-wide (also can be known as
+basic) or client-specific was whether that IBC handler can perform that action
+just using its own context data, which are available upon calling interfaces
+provided by `ClientState` and `ConsensusState` traits, like `is_frozen()`,
+`latest_height()`, etc.
+
+* Latest client state **MUST NOT** be frozen
+* Latest consensus state **MUST** be within the trusting period of the latest
+  client state
```

Comment:
Not necessarily. I think it's more of the protocol bound and looking at it from the security perspective, having that method enforces IBC implementors to consider it at their design stage. Isn't it?

Reply:
I looked up [ICS-2](https://github.com/cosmos/ibc/blob/main/spec/core/ics-002-client-semantics/README.md#updatestate), and it mentions:

> Clients MAY have time-sensitive validity predicates, such that if no ClientMessage is provided for a period of time (e.g. an unbonding period of three weeks) it will no longer be possible to update the client, i.e., the client is being frozen. In this case, a permissioned entity such as a chain governance system or trusted multi-signature MAY be allowed to intervene to unfreeze a frozen client & provide a new correct ClientMessage.

So the trusting period is optional in general. Therefore I think it would be best to have this as a client-specific check? That is, protocols with a trusting period will have that check, but not all protocols have a trusting period.

Reply:
Makes sense. Thanks for clearing this out 
[d59ddfe](https://github.com/cosmos/ibc-rs/pull/383/commits/d59ddfe9c718ded5121c479e562338b9ebcc2c36)

Reply:
I revisited this part. The issue was with the way I classified validations. 
From the implementation standpoint, it does make more sense to categorize validations into two **basic** vs **upgrade-specific** groups. I reverted the previous change: https://github.com/cosmos/ibc-rs/pull/383/commits/ec94b512018aec6c5b8d5947426ccf781734673f